module/reports/helper.js

```diff
@@ -0,0 +1,443 @@
+                    []
+                );
+
+                let tasksReport = {
+                    "total":0,
+                    "completed": 0,
+                    "inProgress": 0,
+                    "notStarted": 0,
+                    "overdue":0
+                };
+                let categories = {
+                    "total":0
+                };
+                let categoriesArray = [];
+                let projectReport = {
+                    "total": 0,
+                    "completed": 0,
+                    "inProgress": 0,
+                    "notStarted": 0,
+                    "overdue": 0,
+                };
+
+                if (!projectDetails.length > 0) {
+
+                    return resolve({
+                        message: CONSTANTS.apiResponses.REPORTS_DATA_NOT_FOUND,
+                        result:  {
+                            data_available:false,
+                            data :{ 
+                                categories: categories,
+                                tasks: tasksReport,
+                                projects: projectReport
+                            }
+                        }
+                    })
+                }
+
+                projectReport["total"] =projectDetails.length;
+
+                   
+                await Promise.all(projectDetails.map(async function (project) {
+
+                    if (project.categories) {
+                        project.categories.map(category => {
+                            if (categoriesArray.includes(category.name)) {
+                                categories[category.name] = categories[category.name] + 1;
+                            } else {
+                                categoriesArray.push(category.name);
+                                categories[category.name] = 1;
+                            }
+                        });
+                        categories['total'] = categories['total'] + project.categories.length;
+                    }
+
+                    if (project.status == CONSTANTS.common.COMPLETED_STATUS) {
+
+                        projectReport[CONSTANTS.common.COMPLETED_STATUS] = projectReport[CONSTANTS.common.COMPLETED_STATUS] + 1;
+
+                    } else if (project.status == CONSTANTS.common.INPROGRESS_STATUS) {
+
+                        var todayDate = moment(project.endDate, "DD.MM.YYYY");
+                        var endDate = moment().format();
+                        if (todayDate.diff(endDate, 'days') < 1) {
+                            projectReport['overdue'] = projectReport['overdue'] + 1;
+                        } else {
+                            projectReport[CONSTANTS.common.INPROGRESS_STATUS] = projectReport[CONSTANTS.common.INPROGRESS_STATUS] + 1;
+                        }
+
+                    } else if (project.status == CONSTANTS.common.NOT_STARTED_STATUS) {
+
+                        var todayDate = moment(project.endDate, "DD.MM.YYYY");
```

Comment:
@rakeshSgr - Define it in one single place

Reply:
resolved